### ISSUE TITLE
Server leak

### ISSUE THREAD

**Issue by eric-j-ason** [borgbackup/borg#6855]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG / ISSUE

## System information. For client/server mode post info for both machines.

Two combinations of server an client tested:

- - Server: Raspberry Pi OS (based on Debian 11) with Borg 1.1.16
   - Client: Ubuntu 20.04.4 with Borg 1.1.15
- - Server: Debian 11 with Borg 1.1.16
   - Client: macOS 11.6.7 with Borg 1.2.1 installed using MacPorts

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

~~~
$ borg info borg@exampe.com:~an_existing_username
Repository path not allowed: /home/an_existing_username
~~~
~~~
$ borg info borg@exampe.com:~a_nonexistent_username
Repository borg@example.com:~a_nonexistent_username does not exist.
~~~

## Describe the problem you're observing.

In case of an existing username, the server expands `~an_existing_username` into that user's home directory and includes it in the error message. In case of a nonexistent username, the server fails to expand it and reports that the repository does not exist.

In both cases, the client gains information about what user accounts are on the server. This is a security concern, as the users on the server might not intend to reveal to the world that they are involved in its operation.

I reckon these are the lines that causes it (taken from current `master` branch):

https://github.com/borgbackup/borg/blob/2418795ed1eebd204357f325efa2304b34696a11/src/borg/remote.py#L328-L331

**Comment by ThomasWaldmann** [comment]
Well, it's not revealed "to the world", but only to other users of that server who are able to login via ssh.

But you are correct that even that might not be wanted. Guess in such cases one could give each user their separated chroot so they only can see their own directories.

**Comment by ThomasWaldmann** [comment]
And guess the only other way to "fix" that would be to replace the error messages by something rather generic in the style of "it did not work" (and not telling any path), which is not disclosing anything, but very unhelpful if you're trying to find out why your own stuff does not work.

**Comment by eric-j-ason** [comment]
> Well, it's not revealed "to the world", but only to other users of that server who are able to login via ssh.

You're right, only those with SSH access. I was imagining a situation where the server hosts backups for a large community, or maybe even customers, who are not personally known to the server operators.

> And guess the only other way to "fix" that would be to replace the error messages by something rather generic in the style of "it did not work" (and not telling any path), which is not disclosing anything, but very unhelpful if you're trying to find out why your own stuff does not work.

Does the `~ ` need to be expanded at all?

Are there other such things that can leak about the server or other clients, that might not be obvious to the involved parties, or is this it?

**Comment by ThomasWaldmann** [comment]
Yes, borg internally needs a filesystem path to the repo. `~` expansion is usually a shell thing (but we do not use a shell here, so we have to expand it at the place you found). fs api calls can't work without it being expanded.

There was no systematic review of all potential leaks yet (IIRC), but other messages and especially python tracebacks come to mind.

But they all share that if you suppress them, debugging some error conditions gets harder or impossible.

**Comment by ThomasWaldmann** [comment]
The repo server (`borg serve`) needs to expand the `~` to the fs path of the home directory. That expanded path is given to normal fs api calls (like `open`).

Strictly taken there is no need to tell the borg client about the exact path, but it can be helpful to know the precise path e.g. if there are permission issues or "file not found" or whatever.

**Comment by eric-j-ason** [comment]
> The repo server (`borg serve`) needs to expand the `~` to the fs path of the home directory.

I guess the part I don't understand is why a `~` ever has to be involved, instead of things working off CWD. Perhaps I'm too locked in to the idea that one [forces a command with SSH](https://borgbackup.readthedocs.io/en/1.2.1/deployment/central-backup-server.html#restrictions):

~~~
command="cd /home/backup/repos/<client fqdn>;
         borg serve --restrict-to-path /home/backup/repos/<client fqdn>",
         restrict <keytype> <key> <host>
~~~

**Comment by ThomasWaldmann** [comment]
> I guess the part I don't understand is why a |~| ever has to be 
> involved, instead of things working off CWD.

Well, borg support both: relative to ~ (home) and relative to . (cwd).

It even supports ~someuser .

> |command="cd /home/backup/repos/<client fqdn>; borg serve 
> --restrict-to-path /home/backup/repos/<client fqdn>", restrict <keytype> 
> <key> <host> |

Guess many people do not have that "cd" there (and it is not required).

Without it, the cwd is usually in the user home usually (not sure if ssh 
guarantees this).

**Comment by ThomasWaldmann** [comment]
Guess for the leakage we have to consider multiple things:

- `~otheruser` is maybe a questionable feature (see abuse documented in top post) and maybe not really needed (the user could always give the full, absolute path)
- `~` expansion is just disclosing **your own** home directory's path (via log messages, maybe tracebacks)
- `.` will also get expanded (via `realpath`) to the full absolute path
- the expansion is technically required. telling the borg client that absolute path is not technically required, but might be useful, esp. in case something does not work as expected.

If we would want to get rid of `~otheruser` expansion (but keep logging, keep `.` and `~` expansion), this could be a breaking change for a few people and releasing the change with borg2 might be a good time.

@enkore @textshell @jdchristensen @m3nu @sophie-h what do you think?

**Comment by ThomasWaldmann** [comment]
Also, guess if we do not want to disclose existing / nonexisting usernames (home directories), guess we need to emit the same error message in both cases, like `Path /home/xxx is not allowed or does not exist`.

**Comment by m3nu** [comment]
In favor of removing `~otheruser`. Not seeing many use cases for this.

I'm not too worried about leaking the absolute path, the less the better of course.

This should also seen in the context of changing to `ssh://`-style paths and making sure we keep supporting what's expected from those.

**Comment by ThomasWaldmann** [comment]
`ssh://` style repos currently do support `/~/...` as well as `/~otheruse/...`.

**Comment by ThomasWaldmann** [comment]
`borg serve --restrict-to-path=... --restrict-to-repository=...` is checked **before** trying to open/create a repository there. I guess we can assume and require that shared hosters caring for security and privacy of their users use at least one of these options to restrict users to their rented place (if not even using chroots or other OS / container / VM level restrictions).

In that case, it will **always** say "not allowed" before even looking what's there / whether there is something at that place.

Thus, I don't think we have to change the log messages.

Because likely few users use it anyway, I removed the `~otheruser` code in #6855.

**Comment by ThomasWaldmann** [comment]
@eric-j-ason fixed by #6864 (in borg branch) and by using the restrict options I mentioned above.

Do you agree?

**Comment by eric-j-ason** [comment]
> @eric-j-ason fixed by #6864 (in borg2 branch) and by using the restrict options I mentioned above.
> 
> Do you agree?

Nice. Thanks!

I agree that not expanding `~otheruser` is the important aspect, so I am satisfied. If I could have my way, I'd would not support `~` at all.

**Comment by ThomasWaldmann** [comment]
OK, fix is merged into borg2 branch (which will likely soon get merged into master).
